x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -0,0 +1,363 @@
+  public update = (
+    sessionId: string,
+    attributes: Partial<BackgroundSessionSavedObjectAttributes>,
+    { savedObjectsClient }: BackgroundSessionDependencies
+  ) => {
+    this.logger.debug(`update | ${sessionId}`);
+    return savedObjectsClient.update<BackgroundSessionSavedObjectAttributes>(
+      BACKGROUND_SESSION_TYPE,
+      sessionId,
+      attributes
+    );
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public delete = (sessionId: string, { savedObjectsClient }: BackgroundSessionDependencies) => {
+    return savedObjectsClient.delete(BACKGROUND_SESSION_TYPE, sessionId);
+  };
+
+  /**
+   * Tracks the given search request/search ID in the saved session (if it exists). Otherwise, just
+   * store it in memory until a saved session exists.
+   * @internal
+   */
+  public trackId = async (
+    searchRequest: IKibanaSearchRequest,
+    searchId: string,
+    { sessionId, isStored }: ISearchOptions,
+    deps: BackgroundSessionDependencies
+  ) => {
+    if (!sessionId || !searchId) return;
+    this.logger.debug(`trackId | ${sessionId} | ${searchId}`);
+    const requestHash = createRequestHash(searchRequest.params);
+
+    // If there is already a saved object for this session, update it to include this request/ID.
+    // Otherwise, just update the in-memory mapping for this session for when the session is saved.
+    if (isStored) {
+      const attributes = { idMapping: { [requestHash]: searchId } };
+      await this.update(sessionId, attributes, deps);
+    } else {
+      const map = this.sessionSearchMap.get(sessionId) ?? {
+        insertTime: moment(),
+        retryCount: 0,
+        ids: new Map<string, string>(),
+      };
+      map.ids.set(requestHash, searchId);
+      this.sessionSearchMap.set(sessionId, map);
+    }
+  };
+
+  /**
+   * Look up an existing search ID that matches the given request in the given session so that the
+   * request can continue rather than restart.
+   * @internal
+   */
+  public getId = async (
+    searchRequest: IKibanaSearchRequest,
+    { sessionId, isStored, isRestore }: ISearchOptions,
+    deps: BackgroundSessionDependencies
+  ) => {
+    if (!sessionId) {
+      throw new Error('Session ID is required');
+    } else if (!isStored) {
+      throw new Error('Cannot get search ID from a session that is not stored');
+    } else if (!isRestore) {
+      throw new Error('Get search ID is only supported when restoring a session');
+    }
+
+    const session = await this.get(sessionId, deps);
+    const requestHash = createRequestHash(searchRequest.params);
+    if (!session.attributes.idMapping.hasOwnProperty(requestHash)) {
+      throw new Error('No search ID in this session matching the given search request');
+    }
+
+    return session.attributes.idMapping[requestHash];
+  };
+
+  public asScopedProvider = ({ savedObjects }: CoreStart) => {
+    return (request: KibanaRequest) => {
```

Comment:
👍 Thanks for introducing the scoped provider.

Since we haven't granted any privileges for this new saved object type, these requests will all fail our authorization checks unless the user is a superuser.

I know we're discussing background session security outside of this PR, but I wanted to raise it anyway in case this wasn't your expectation.

Reply:
So how should this be correctly addressed?

Reply:
We'll either take the sub-feature privilege approach, or the top-level feature approach that we've been discussing in the doc that Brandon put together. Once we reach consensus I'll help you with the changes that we'll need to make